### DISCUSSION TITLE
How to add spots during different phases of the cycle

### DISCUSSION THREAD

**Discussion by 0batsy0** [phoebe-project/phoebe2#877] [Questions]
We have three complete light curves observed over different time periods, with only the second segment showing the O'Connell effect. Therefore, we aim to use a single set of parameters for fitting all segments simultaneously, but we want to add spots only to the second segment of the data. We are unsure how to achieve this.

**Comment by kecnry** [comment]
Unfortunately for your use-case, the `enabled` toggle on spots (features) is per-compute but not per-dataset.  If you're only running forward models, then you can toggle this manually and swap out your datasets as needed, but it sounds like you're trying to use phoebe's fitting capability and so need to handle this within a single call to the forward model.  In theory, the following should work:

```
import phoebe

b = phoebe.default_binary()
b.add_spot(component='primary', relteff=0.8, feature='spot01')

# add multiple datasets, each dataset will either have the spot enabled or disabled via separate compute options
# for this example, we'll just add one of each
b.add_dataset('lc', compute_phases=phoebe.linspace(0,1,101), dataset='lc_with_spot')
b.add_dataset('lc', compute_phases=phoebe.linspace(0,1,101), dataset='lc_without_spot')

# now we'll create separate compute options: one which will handle all datasets with the spot enabled and one
# for all datasets with the spot disabled
b.add_compute('phoebe', compute='comp_with_spot')
b.set_value('enabled', dataset='lc_without_spot', compute='comp_with_spot', value=False)

b.add_compute('phoebe', compute='comp_without_spot')
b.set_value('enabled', dataset='lc_with_spot', compute='comp_without_spot', value=False)
b.set_value('enabled', feature='spot01', compute='comp_without_spot', value=False)

b.run_compute(compute=['comp_with_spot', 'comp_without_spot'])
```

I ran into a small bug running this which is fixed in #878 and will be included in a bugfix release in the next week or so.  Similarly to #839, this is probably also susceptible to the plotting bug reported as #840.

However, in order to use this for fitting, phoebe will need to be extended to support multiple computes being passed to the solver backends.  This is not a case we had considered when designing the fitting infrastructure, but I can see how it would be useful and don't think there is anything that would make it infeasible.  I created #879 to track that effort, but with a backlog of higher-priority bugs, I can't guarantee exactly when that will be done.  Should you or anyone be interested in trying to implement this in phoebe, I would be happy to give pointers to get started.  

In the meantime, you should be able to do the above for now (after the next bugfix release, or just by calling `run_compute` twice with one compute option each time) though and use external fitting.

If you would rather use phoebe's internal fitting functionality and are willing to pay computational overhead, it _should_ be possible to make use of the [custom cost function capability](http://phoebe-project.org/docs/2.4/tutorials/emcee_custom_lnprob) to effectively _re_-compute some of the datasets before evaluating the cost function.  If you want to go in this direction and get stuck, let me know and I can try to put together a minimal example of how that might look.

**Reply by 0batsy0** [reply]
Thank you for your response, but I still have some questions. For example, if the three segments of data have different spots, such as relteff=0.5 for the first segment and relteff=0.8 for the second segment, I want to add different spots for different periods and perform MCMC calculations on all the data. This seems difficult to achieve.

**Reply by kecnry** [reply]
Yes, that is definitely an advanced use-case and somewhat breaks the concept of a single physical model (unless/until we implement time-dependent parameters so that spots could migrate in time).  You should be able to extend the example above to any number of spots and datasets, you'll just need to toggle `enabled` on the the spots and datasets either in separate compute options or by separate calls to `run_compute`.  Have you considered accounting for the spots with gaussian processes instead?
